### PR TITLE
[Upstream _NamedOptimzer] Reland PR (89480)

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Reland https://github.com/pytorch/pytorch/pull/89480/
* #90294
* __->__ #90293

Differential Revision: [D41783015](https://our.internmc.facebook.com/intern/diff/D41783015)

### CODE REVIEW

**Comment by fduwjj** [comment]
@pytorch rebase

File: test/distributed/optim/test_named_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            torch.optim.SGD,
+            lr=1e-2,
+            momentum=0.6,
+        )
+
+        err_msg = (
+            "Expects the optim to be initialized before load but found not initialized"
+        )
+        with self.assertRaisesRegex(ValueError, err_msg):
+            named_optim_2.load_state_dict(state_dict_to_load)
```
Comment (awgu) [inline test/distributed/optim/test_named_optimizer.py:245]: nit: Add newline (might be required by linter).

**Review by awgu — APPROVED** [review]
I am stamping since I think this is the same as before just without `__all__`.

Please make sure `python test/test_public_bindings.py` passes and that the lint passes before merging. Thanks!

File: test/distributed/optim/test_named_optimizer.py (hunk 1 above)
Comment (fduwjj) [reply test/distributed/optim/test_named_optimizer.py:245]: Linter does not complain previously. But adding it now.

**Comment by fduwjj** [comment]
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41783015).

**Pull request closed without merging**